projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py

```diff
@@ -0,0 +1,284 @@
+                consequence="Your credentials are not refreshed and VDK control operations that require authentication "
+                "will not work.",
+                countermeasure="Check error message and follow instructions in it. Check your network connectivity."
+                " Check if you can access the Oauth2 server. Retry to login a few times again."
+                " Contact SRE Team operating Oauth2 server for help if nothing else works.",
+            ) from e
+        # Response per spec in https://tools.ietf.org/html/rfc6750#section-4
+        # see also https://www.oauth.com/oauth2-servers/access-tokens/access-token-response
+        self.update_access_token(
+            token_response[AuthRequestValues.ACCESS_TOKEN_KEY.value]
+        )
+        self.update_access_token_expiration_time(
+            time.time()
+            + int(token_response.get(AuthRequestValues.EXPIRATION_TIME_KEY.value, "0"))
+        )
+
+    def read_access_token(self) -> Optional[str]:
+        """
+        Read access token from _cache or fetch it from Authorization server.
+        If not available in _cache it will get it using provided configuration during VDK CLI login to fetch it.
+        If it detects that token is about to expire it will try to refresh it.
+        :return: the access token or None if it cannot detect any credentials.
+        """
+        if (
+            not self._cache.access_token
+            or self._cache.access_token_expiration_time < time.time() + 60
+        ):
+            log.debug("Acquire access token (it's either expired or missing) ...")
+            self.acquire_and_cache_access_token()
+        return self._cache.access_token
+
+    def acquire_and_cache_access_token(self):
+        """
+        Acquires and caches access token
+        """
+        log.debug(f"Using auth type {self._cache.auth_type} to acquire access token")
+        if (
+            self._cache.auth_type == LoginTypes.API_TOKEN.value
+            and self._configured_api_token()
+        ):
+            self.__exchange_api_for_access_token()
+        elif (
+            self._cache.auth_type == LoginTypes.CREDENTIALS.value
+            and self._configured_refresh_token()
+        ):
+            self.__exchange_refresh_for_access_token()
+        elif (
+            self.__vdk_config.api_token
+            and self.__vdk_config.api_token_authorization_url
+        ):
+            self.update_api_token(self.__vdk_config.api_token)
+            self.update_api_token_authorization_url(
+                self.__vdk_config.api_token_authorization_url
+            )
+            self.__exchange_api_for_access_token()
+        else:
+            log.debug(
+                "No authentication mechanism found. Will not cache access token."
+                "If Control Service authentication is enabled, API calls will fail."
+            )
+
+    def __exchange_refresh_for_access_token(self):
```

Comment:
nit but maybe move __exchange_refresh_for_access_token up so it's next to __exchange_api_for_access_token

Reply:
Moved. Thanks